File: client/site_tests/platform_MiniJailPidNamespace/platform_MiniJailPidNamespace.py
# Copyright (c) 2010 The Chromium OS Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import logging
import os
import re

from autotest_lib.client.bin import test
from autotest_lib.client.common_lib import error, utils

class platform_MiniJailPidNamespace(test.test):
    version = 1
    preserve_srcdir = True

    def setup(self):
        os.chdir(self.srcdir)
        utils.system('make clean')
        utils.system('make all')


    def __run_cmd(self, cmd):
        result = utils.system_output(cmd, retain_output=True,
                                     ignore_status=True)
        return result


    def run_once(self):
        # Check that --namespace-pid works
        check_cmd = os.path.join(self.bindir, 'platform_MiniJailPidNamespace')
        cmd = ('/sbin/minijail --namespace-pid -- ' + check_cmd)
        result = self.__run_cmd(cmd)
        succeed_pattern = re.compile(r"SUCCEED: (.+)")
        success = succeed_pattern.findall(result)
        if len(success) == 0:
          raise error.TestFail('Set PID namespace failed.')

File: client/site_tests/platform_MiniJailPidNamespace/src/platform_MiniJailPidNamespace.cc
// Copyright (c) 2010 The Chromium OS Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>

// Checks that getpid() returns 1.  If this process has its own PID namespace,
// then it should have a PID of 1.
void CheckProcessIdIsOne() {
  pid_t pid;
  if ((pid = getpid()) == (static_cast<pid_t>(1))) {
    printf("SUCCEED: Process ID is 1\n");
  } else {
    printf("FAIL: Process ID is %d\n", pid);
  }
}

int main(int argc, const char* argv[]) {
  CheckProcessIdIsOne();

  return 0;
}
